commons/src/lexer/token/iterator/slice.rs

```diff
@@ -0,0 +1,137 @@
+use itertools::PeekingNext;
+
+use crate::lexer::token::{Token, TokenKind};
+
+#[derive(Debug, Default, Clone)]
+pub struct TokenSliceIterator<'slice, 'input> {
+    /// The [`Symbol`] slice the iterator was created for.
+    tokens: &'slice [Token<'input>],
+    /// The current index of the iterator inside the [`Symbol`] slice.
+    index: usize,
+    /// The peek index of the iterator inside the [`Symbol`] slice.
+    peek_index: usize,
+    scope: usize,
+}
+
+impl<'slice, 'input, T> From<T> for TokenSliceIterator<'slice, 'input>
+where
+    T: Into<&'slice [Token<'input>]>,
+{
+    fn from(value: T) -> Self {
+        TokenSliceIterator {
+            tokens: value.into(),
+            index: 0,
+            peek_index: 0,
+            scope: 0,
+        }
+    }
+}
+
+impl<'slice, 'input> Iterator for TokenSliceIterator<'slice, 'input> {
+    type Item = &'slice Token<'input>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let token = self.tokens.get(self.index)?;
+
+        self.index += 1;
+        self.peek_index = self.index;
+
+        Some(token)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (0, Some(self.max_len()))
+    }
+}
+
+impl<'slice, 'input> PeekingNext for TokenSliceIterator<'slice, 'input> {
+    fn peeking_next<F>(&mut self, accept: F) -> Option<Self::Item>
+    where
+        Self: Sized,
+        F: FnOnce(&Self::Item) -> bool,
+    {
+        let token = self.tokens.get(self.peek_index).filter(accept)?;
+        self.peek_index += 1;
+        Some(token)
+    }
+}
+
+impl<'slice, 'input> TokenSliceIterator<'slice, 'input> {
+    /// Returns the maximum length of the remaining [`Symbol`]s this iterator might return.
+    ///
+    /// **Note:** This length does not consider parent iterators, or matching functions.
+    /// Therefore, the returned number of [`Symbol`]s might differ, but cannot be larger than this length.
+    pub fn max_len(&self) -> usize {
+        if self.tokens.is_empty() {
+            return 0;
+        }
+
+        self.tokens[self.index.min(self.tokens.len() - 1)..].len()
```

Comment:
This and other similar functions can be implemented as: 

```suggestion
        self.tokens.len().saturating_sub(self.index)
```